exchanges/bybit/bybit.go

```diff
@@ -0,0 +1,729 @@
+func (by *Bybit) GetWalletBalance(ctx context.Context) ([]Balance, error) {
+	resp := struct {
+		Data struct {
+			Balances []Balance `json:"balances"`
+		} `json:"result"`
+	}{}
+	err := by.SendAuthHTTPRequest(ctx, exchange.RestSpot, http.MethodGet, bybitWalletBalance, url.Values{}, &resp, privateSpotRate)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Data.Balances, nil
+}
+
+// SendHTTPRequest sends an unauthenticated request
+func (by *Bybit) SendHTTPRequest(ctx context.Context, ePath exchange.URL, path string, f request.EndpointLimit, result interface{}) error {
+	endpointPath, err := by.API.Endpoints.GetURL(ePath)
+	if err != nil {
+		return err
+	}
+	item := &request.Item{
+		Method:        http.MethodGet,
+		Path:          endpointPath + path,
+		Result:        result,
+		Verbose:       by.Verbose,
+		HTTPDebugging: by.HTTPDebugging,
+		HTTPRecording: by.HTTPRecording}
+
+	return by.SendPayload(ctx, f, func() (*request.Item, error) {
+		return item, nil
+	})
+}
+
+// SendAuthHTTPRequest sends an authenticated HTTP request
+func (by *Bybit) SendAuthHTTPRequest(ctx context.Context, ePath exchange.URL, method, path string, params url.Values, result interface{}, f request.EndpointLimit) error {
+	if !by.AllowAuthenticatedRequest() {
+		return fmt.Errorf("%s %w", by.Name, exchange.ErrAuthenticatedRequestWithoutCredentialsSet)
+	}
+	endpointPath, err := by.API.Endpoints.GetURL(ePath)
+	if err != nil {
+		return err
+	}
+
+	if params == nil {
+		params = url.Values{}
+	}
+
+	if params.Get("recvWindow") == "" {
+		params.Set("recvWindow", strconv.FormatInt(defaultRecvWindow.Milliseconds(), 10))
+	}
+
+	return by.SendPayload(ctx, f, func() (*request.Item, error) {
+		params.Set("timestamp", strconv.FormatInt(time.Now().UnixMilli(), 10))
+		params.Set("api_key", by.API.Credentials.Key)
+		signature := params.Encode()
+		hmacSigned, err := crypto.GetHMAC(crypto.HashSHA256, []byte(signature), []byte(by.API.Credentials.Secret))
+		if err != nil {
+			return nil, err
+		}
+		hmacSignedStr := crypto.HexEncodeToString(hmacSigned)
+
+		headers := make(map[string]string)
+		var payload []byte
+		switch method {
+		case http.MethodPost:
+			headers["Content-Type"] = "application/json"
+			m := make(map[string]string)
+			m["api_key"] = by.API.Credentials.Key
+
+			for k, v := range params {
+				m[k] = strings.Join(v, "")
+			}
+			m["sign"] = hmacSignedStr
+			payload, err = json.Marshal(m)
+			if err != nil {
+				return nil, err
+			}
+		default:
+			headers["Content-Type"] = "application/x-www-form-urlencoded"
+			path = common.EncodeURLValues(path, params)
+			path += "&sign=" + hmacSignedStr
+		}
+		return &request.Item{
+			Method:        method,
+			Path:          endpointPath + path,
+			Headers:       headers,
+			Body:          bytes.NewBuffer(payload),
+			Result:        &result,
+			AuthRequest:   true,
+			Verbose:       by.Verbose,
+			HTTPDebugging: by.HTTPDebugging,
+			HTTPRecording: by.HTTPRecording}, nil
+	})
```

Comment:
It would be good to also add in error capturing to the response. See `exchanges/binance/binance.go`'s `SendAuthHTTPRequest` `errCap` work as that will interpret the extra response data from binance. In ByBit's case, it would be to parse `"ret_code"`,`"ret_msg"`, `"ext_code"`,`"ext_info"` fields and look for errors and error messages so it can be passed to the user

For example, setting my API key to "blahblah" returns the following:
```
HTTP status: 200 OK, Code: 200
Bybit raw response: {"ret_code":-1022,"ret_msg":"Signature for this request is not valid.","ext_code":null,"ext_info":null,"result":null}
```
So Bybit returns 200 status messages for failure and you need to parse the response to find out any errors. This also means that current auth tests can pass even when they fail